### ISSUE TITLE
Execution variables cannot be used in spark_event_logs_s3_uri

### ISSUE THREAD

**Issue by nebur395** [aws/sagemaker-python-sdk#3195]
**Describe the bug**
In order to be able to organise our S3 bucket in which the input and the outputs are being uploaded for our Sagemaker pipelines, we want to be able to specify a custom `spark_event_logs_s3_uri` path. To improve our data lineage we are trying to store these files by pipeline name and execution ids. 

Therefore we need to use ExecutionVariables in `spark_event_logs_s3_uri` property. For using them, you have to use the special `Join` function from sagemaker, otherwise you cannot concatenate them with other static strings. So final output for this property would be:
```
spark_event_logs_s3_uri=Join(on="/", values=[
    "s3:/", default_bucket, ExecutionVariables.PIPELINE_NAME, ExecutionVariables.PIPELINE_EXECUTION_ID, "spark-event-logs/"
])
```

With that said, we are getting an error which says: `AttributeError: 'Join' object has no attribute 'decode'`. However, if you just remove the Join and the Execution variables for something like this: `spark_event_logs_s3_uri=f"s3://{default_bucket}/spark-event-logs/"` then it works.

**To reproduce**
```
import boto3
import sagemaker
from sagemaker.workflow.pipeline import Pipeline
from sagemaker.processing import ProcessingInput, ProcessingOutput
from sagemaker.workflow.steps import ProcessingStep
from sagemaker.spark.processing import PySparkProcessor
from sagemaker.workflow.execution_variables import ExecutionVariables
from sagemaker.workflow.functions import Join

boto_session = boto3.Session(region_name="eu-west-1")

sagemaker_client = boto_session.client("sagemaker")

default_bucket = "XYZ"

sagemaker_session = sagemaker.session.Session(
    boto_session=boto_session,
    sagemaker_client=sagemaker_client,
    default_bucket=default_bucket
)

bucket_pipeline_execution_prefix = [
    "s3:/", default_bucket, ExecutionVariables.PIPELINE_NAME, ExecutionVariables.PIPELINE_EXECUTION_ID
]

spark_processor = PySparkProcessor(
    base_job_name="sm-spark",
    role="XYZ",
    instance_count=1,
    instance_type="ml.m5.xlarge",
    max_runtime_in_seconds=1200,
    image_uri="XYZ",
    sagemaker_session=sagemaker_session
)

run_args = spark_processor.get_run_args(
    submit_app="./processing.py",
    inputs=[],
    outputs=[
        ProcessingOutput(
            source='/opt/ml/processing/output',
            output_name='demo-output',
            destination=Join(on="/", values=[
                *bucket_pipeline_execution_prefix, "output/preprocessed"
            ])
        )
    ],
    spark_event_logs_s3_uri=Join(on="/", values=[
                *bucket_pipeline_execution_prefix, "spark-event-logs"
            ])
)

step_process = ProcessingStep(
    name="demo-processing",
    processor=spark_processor,
    inputs=run_args.inputs,
    outputs=run_args.outputs,
    code=run_args.code,
)

pipeline = Pipeline(
    name='demo-pipeline',
    parameters=[],
    sagemaker_session=sagemaker_session,
    steps=[step_process],
)

pipeline.upsert(role_arn="XYZ")

execution = pipeline.start()
```

**Expected behavior**
Script to work and spark logs to be stored in the same way as `step_preprocess.properties.ProcessingOutputConfig.Outputs["demo-output"].S3Output.S3Uri`

**Screenshots or logs**
```
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-21-cf9b6fbcda40> in <module>
     55     ],
     56     spark_event_logs_s3_uri=Join(on="/", values=[
---> 57                 "s3:/", default_bucket, ExecutionVariables.PIPELINE_NAME, ExecutionVariables.PIPELINE_EXECUTION_ID, "spark-event-logs/"
     58             ])
     59 )

/opt/conda/lib/python3.7/site-packages/sagemaker/spark/processing.py in get_run_args(self, submit_app, submit_py_files, submit_jars, submit_files, inputs, outputs, arguments, job_name, configuration, spark_event_logs_s3_uri)
    784             submit_files=submit_files,
    785             configuration=configuration,
--> 786             spark_event_logs_s3_uri=spark_event_logs_s3_uri,
    787         )
    788 

/opt/conda/lib/python3.7/site-packages/sagemaker/spark/processing.py in _extend_processing_args(self, inputs, outputs, **kwargs)
    900         )
    901 
--> 902         return super()._extend_processing_args(extended_inputs, outputs, **kwargs)
    903 
    904 

/opt/conda/lib/python3.7/site-packages/sagemaker/spark/processing.py in _extend_processing_args(self, inputs, outputs, **kwargs)
    267         if kwargs.get("spark_event_logs_s3_uri"):
    268             spark_event_logs_s3_uri = kwargs.get("spark_event_logs_s3_uri")
--> 269             self._validate_s3_uri(spark_event_logs_s3_uri)
    270 
    271             self._spark_event_logs_s3_uri = spark_event_logs_s3_uri

/opt/conda/lib/python3.7/site-packages/sagemaker/spark/processing.py in _validate_s3_uri(self, spark_output_s3_path)
    593             spark_output_s3_path (str): The URI of the Spark output S3 Path.
    594         """
--> 595         if urlparse(spark_output_s3_path).scheme != "s3":
    596             raise ValueError(
    597                 f"Invalid s3 path: {spark_output_s3_path}. Please enter something like "

/opt/conda/lib/python3.7/urllib/parse.py in urlparse(url, scheme, allow_fragments)
    365     Note that we don't break the components up in smaller bits
    366     (e.g. netloc is a single string) and we don't expand % escapes."""
--> 367     url, scheme, _coerce_result = _coerce_args(url, scheme)
    368     splitresult = urlsplit(url, scheme, allow_fragments)
    369     scheme, netloc, url, query, fragment = splitresult

/opt/conda/lib/python3.7/urllib/parse.py in _coerce_args(*args)
    121     if str_input:
    122         return args + (_noop,)
--> 123     return _decode_args(args) + (_encode_result,)
    124 
    125 # Result objects are more helpful than simple tuples

/opt/conda/lib/python3.7/urllib/parse.py in _decode_args(args, encoding, errors)
    105 def _decode_args(args, encoding=_implicit_encoding,
    106                        errors=_implicit_errors):
--> 107     return tuple(x.decode(encoding, errors) if x else '' for x in args)
    108 
    109 def _coerce_args(*args):

/opt/conda/lib/python3.7/urllib/parse.py in <genexpr>(.0)
    105 def _decode_args(args, encoding=_implicit_encoding,
    106                        errors=_implicit_errors):
--> 107     return tuple(x.decode(encoding, errors) if x else '' for x in args)
    108 
    109 def _coerce_args(*args):

AttributeError: 'Join' object has no attribute 'decode'
```

**System information**
Sagemaker studio notebook:
- Instance type: ml.t3.medium
- Notebook environment: 
  - Image: Data science
  - Kernel: Python 3

**Comment by qidewenwhen** [comment]
Hi @nebur395, thanks for reaching out!
This pending PR should fix this issue: https://github.com/aws/sagemaker-python-sdk/pull/3167/

**Comment by qidewenwhen** [comment]
The PR has been merged recently. Will be released in a week.

**Comment by qidewenwhen** [comment]
Hi, the fix has been released in v2.100.0: https://github.com/aws/sagemaker-python-sdk/releases/tag/v2.100.0
Can you confirm if it works for you?

**Comment by nebur395** [comment]
Hey @qidewenwhen, thanks a lot for the support really. I've just checked it and it works smoothly with sagemaker@2.100.0, so I close the issue.
